hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1612 @@
+    #[inline]
+    fn set_rcfg(&mut self, cfg: Rcfg) {
+        self.rcfg = cfg;
+    }
+    /// Reset [`Rcfg`] to default values
+    #[inline]
+    fn reset_rcfg_to_default(&mut self) {
+        self.rcfg = Rcfg::default();
+    }
+    /// Set [`Rnext`]
+    #[inline]
+    fn set_rnext(&mut self, next: Rnext) {
+        self.rnext = next;
+    }
+}
+
+impl IcmSecondaryRegionDesc {
+    pub const fn default() -> Self {
+        IcmSecondaryRegionDesc {
+            raddr: Raddr::default(),
+            rcfg: Rcfg::default(),
+            _pad: 0,
+            rnext: Rnext::default(),
+        }
+    }
+}
+
+/// ICM Hash Area
+///
+/// Follows C-structure conventions and is 128-byte aligned
+///
+/// >**Important**
+/// >
+/// >ICM engine will **read** or **write** to wherever this
+/// >is instantiated in memory, based on the [`Icm::set_hash_addr()`]
+/// >so the user must ensure that this variable lives long enough or is
+/// >static
+#[derive(Debug)]
+#[repr(C)]
+#[repr(align(128))]
+pub struct IcmHashArea {
+    pub region0: [u32; 8],
+    pub region1: [u32; 8],
+    pub region2: [u32; 8],
+    pub region3: [u32; 8],
+}
+
+impl IcmHashArea {
+    pub fn region0(&self) -> [u32; 8] {
+        self.region0
+    }
+    pub fn region1(&self) -> [u32; 8] {
+        self.region1
+    }
+    pub fn region2(&self) -> [u32; 8] {
+        self.region2
+    }
+    pub fn region3(&self) -> [u32; 8] {
+        self.region3
+    }
+    pub const fn default() -> Self {
+        IcmHashArea {
+            region0: [0; 8],
+            region1: [0; 8],
+            region2: [0; 8],
+            region3: [0; 8],
+        }
+    }
+}
+
+/// Region Start Address Structure
+///
+/// Follows C-structure conventions
+#[derive(Clone, Copy)]
+#[repr(C)]
+pub struct Raddr {
+    pub raddr: u32,
+}
+
+impl Raddr {
+    pub fn set_raddr<T>(&mut self, raddr: *const T) {
+        self.raddr = raddr as u32;
+    }
+    const fn default() -> Self {
+        // Unsure what a good Raddr default should be,
+        // this is at least easily spotted when debugging...
+        Raddr {
+            raddr: 0xDEADBEEF_u32,
+        }
+    }
+}
+
+/// Region Configuration Structure
+///
+/// Follows C-structure conventions
+#[derive(Clone, Copy)]
+#[repr(C)]
+pub struct Rcfg {
+    rcfg: u32,
+}
+
+impl Rcfg {
+    const fn default() -> Self {
+        Rcfg { rcfg: 0x3F0 }
+    }
+}
+
+impl Rcfg {
+    /// Compare Digest or Write Back Digest
+    ///
+    /// * true: in Compare Digest mode.
+    /// * false: in Write Back Digest mode.
+    #[inline]
+    pub fn set_cdwbn(&mut self, value: bool) {
+        self.rcfg = (self.rcfg & !0x01) | (value as u32);
+    }
+    /// Wrap Command
+    ///
+    /// * true: next region descriptor address loaded is `DSCR`.
+    /// * false: the next region descriptor address is `current + 0x10`.
+    #[inline]
+    pub fn set_wrap(&mut self, value: bool) {
+        self.rcfg = (self.rcfg & !(0x01 << 1)) | ((value as u32) << 1);
+    }
```

Comment:
This looks like a lot of manual work to maintain a set of bit fields. I think you should consider using `modular_bitfields` or a similar crate. I know @jbeaurivage made good use of it in the DMA module for his descriptors.

Reply:
Once again, great idea. Simplified it quite a bit.